src/pvcam.cpp

```diff
@@ -0,0 +1,1279 @@
+#include "pvcam.h"
+
+#define L (aq_logger)
+#define LOG(...) L(0, __FILE__, __LINE__, __FUNCTION__, __VA_ARGS__)
+#define LOGE(...) L(1, __FILE__, __LINE__, __FUNCTION__, __VA_ARGS__)
+
+// #define TRACE(...) LOG(__VA_ARGS__)
+#define TRACE(...)
+
+#define EXPECT(e, ...)                                                         \
+    do {                                                                       \
+        if (!(e)) {                                                            \
+            LOGE(__VA_ARGS__);                                                 \
+            throw std::runtime_error("Expression was false: " #e);             \
+        }                                                                      \
+    } while (0)
+#define CHECK(e) EXPECT(e, "Expression was false:\n\t%s\n", #e)
+
+#define PVCAM_INNER(e, mtx, logger, action)                                    \
+    do {                                                                       \
+        std::scoped_lock<std::mutex> lock((mtx));                              \
+        if (PV_OK != (e)) {                                                    \
+            const int16 code = pl_error_code();                                \
+            char msg[ERROR_MSG_LEN];                                           \
+            pl_error_message(code, msg);                                       \
+            logger(msg);                                                       \
+            action;                                                            \
+        }                                                                      \
+    } while (0)
+
+#define PVCAM(e, mtx)                                                          \
+    PVCAM_INNER(                                                               \
+      e, mtx, LOGE, throw std::runtime_error("PVCAM API call failed."))
+#define PVWARN(e, mtx) PVCAM_INNER(e, mtx, LOG, )
+
+namespace {
+/// Utilities
+
+// Maps Acquire SampleType to GenICam pixel format strings.
+const std::unordered_map<SampleType, PL_IMAGE_FORMATS>
+  sample_type_to_pixel_format{
+      { SampleType_u8, PL_IMAGE_FORMAT_MONO8 },
+      { SampleType_u16, PL_IMAGE_FORMAT_MONO16 },
+  };
+
+const std::unordered_map<PL_IMAGE_FORMATS, SampleType>
+  pixel_format_to_sample_type{
+      { PL_IMAGE_FORMAT_MONO8, SampleType_u8 },
+      { PL_IMAGE_FORMAT_MONO16, SampleType_u16 },
+  };
+
+/**
+ * @brief Checks that the given camera id is within the range of a signed
+ * 16-bit.
+ * @param id The camera id to check.
+ * @throws std::runtime_error if the id is out of range.
+ */
+void
+check_pvcam_camera_id(uint64_t id)
+{
+    constexpr uint64_t limit = std::numeric_limits<int16>::max();
+    EXPECT(id <= limit, "Expected a value at most %llu. Got %llu.", limit, id);
+}
+
+void
+camera_properties_to_rgn_type(CameraProperties* properties, rgn_type& roi)
+{
+    auto binning = (uns16)properties->binning;
```

Comment:
Yeah, it comes from the PVCAM headers. I went with using the data types used by the PVCAM headers when I expect them to be used with / returned from PVCAM API calls. I can see how it would be confusing, and it's almost certain they won't change their typedefs, so I can change them to `u?int\d\d_t` types.